### DISCUSSION TITLE
Using a sprite as character map for a smooth font and display them using colors

### DISCUSSION THREAD

**Discussion by delhoume** [Bodmer/TFT_eSPI#1123] [New processor/display support or capability request]
Hi,

I have an application (a matrix rain effect) where I display a limited number of chars (about 100), but with different colors.
My application flickers somehow (I created a sprite for each vertical line to reduce this) but I have to draw each char in its final color to the vertical sprite, then push the sprite to the screen.
I would like to create a sprite that contains all chars (in grey levels), and display them to the screen "tinted".
Creating the sprite and filling it with chars from the smooth font is easy but I did not find a way to display, for a given char, the correct sprite sub-section tinted to the final screen.
There are many colors so creating a sprite with all combinations of chars and colors is not possible.

That looks like some kind of cache for smooth fonts chars, instead of parsing the vlw each time for  a given char, its contents would be in a sprite, and displayed to the final destination using a tint color.

The background color is uniform (black).

I am clear enough ?

Current code for the application is at https://github.com/delhoume/MatrixRain

**Comment by Bodmer** [comment]
I tried the sketch but it would not compile under the Arduino environment.

**Reply by delhoume** [reply]
I am using Arduino IDE version  1.8.13 with a Wemos D1 mini (esp8266) connected to the TFT screen.
The sketch has only a dependency to local DigitalRainAnim.<h,cpp> and TFT_eSPI.

The code is quite simple, rendering vertical lines, each composed of a number of characters.

What is the error you get ?

**Reply by delhoume** [reply]
Got it, you can update now, I moved the Processing sample code to its own folder, as it was compiled by Arduino...

**Reply by Bodmer** [reply]
I don't have an ESP8266 setup at the moment so used and ESP32 and still get errors.
```
DigitalRainAnim.h:43:10: error: 'DigitalRainAnim::line_speed' will be initialized after [-Werror=reorder]
     int* line_speed;   //dynamic array for eacg line speed
          ^
DigitalRainAnim.h:29:18: error:   'TFT_eSprite* DigitalRainAnim::spr' [-Werror=reorder]
     TFT_eSprite* spr; 
                  ^
DigitalRainAnim.cpp:9:1: error:   when initialized here [-Werror=reorder]
 DigitalRainAnim::DigitalRainAnim() 
 ^
```

**Reply by delhoume** [reply]
Using my ESP32 env 1.0.4 and ESP32 Wrover module as board it compiles, I would say that your environment is picky about order of initializations in the constructor, so I updated again to put spr first and match .h order, you can update

**Comment by Bodmer** [comment]
I had a look, most of the screen update time is due to  the poor random access time of SPIFFS which caches blocks in RAM and ends up flushing and reloading.

Convert the font to an array, [see examples here](https://github.com/Bodmer/TFT_eSPI/blob/master/examples/Smooth%20Fonts/FLASH_Array/Font_Demo_1_Array/NotoSansBold15.h).

On an ST7796 320x240 display clocked at 80MHz SPI rate this boosted performance from 8fps to 30fps.

**Reply by delhoume** [reply]
Thanks for the hint !
I updated the code to use font array and removed SPIFFs dependencies
